### PR TITLE
Add ETF support

### PR DESCRIPTION
**Pull request by MinnDevelopment**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

This adds a custom implementation to parse and serialize ETF payloads for the discord gateway connection.

You can enable this by using `JDABuilder#setGatewayEncoding(GatewayEncoding.ETF)`, this will be an opt-in.

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/JDABuilder.java [hunk 1]
```diff
@@ -530,6 +531,13 @@ private JDABuilder setDisabledCache(EnumSet<CacheFlag> flags)
         return this;
     }
 
+    @Nonnull
+    public JDABuilder setGatewayEncoding(@Nonnull GatewayEncoding encoding)
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/JDABuilder.java:535]: This also needs to be added to the shard manager probably

File: src/main/java/net/dv8tion/jda/api/utils/data/etf/ExTermEncoder.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    {
+        buffer = realloc(buffer, 5);
+        buffer.put(INT);
+        buffer.putInt(value);
+        return buffer;
+    }
+
+    private static ByteBuffer packLong(ByteBuffer buffer, long value)
+    {
+        byte sign = (byte) (value < 0 ? 1 : 0);
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/utils/data/etf/ExTermEncoder.java:116]: I'm not sure if we should just treat all longs here as unsigned, discord doesn't really need signed longs for anything.

**Comment by MinnDevelopment** [comment]
@napstr gave me some performance insights. Using ETF instead of JSON seems to have almost no performance impact. This is most likely due to the decompression time which should far outweigh the parsing time. Regardless, we can add support for it as an option since the code works fine and having fallbacks is always good.

**Comment by napstr** [comment]
Adding some more details: Without doing in depth profiling, general JVM metrics like CPU & memory usage and GC activity look fairly similar. Unsure if the slight decrease in CPU compared to -7d can be attributed to the ETF change, or it just being a fresh start. A snapshot of the data can be found here: https://grafana.npstr.space/dashboard/snapshot/p6Myi0ARKjW4uTBk9mBWggga0focaNnT

**Pull request merged**